### PR TITLE
[HDFS-489] Consume rack info by HDFS

### PR DESCRIPTION
**Pull request by NimaVaziri**
This adds a few endpoints to the scheduler for getting zone information based on task name and ip address (as needed by way of HDFS functionality).

It also moves the `ZoneValidator` validator class to a generic scheduler location as it can be used by multiple services.

Verification of functionality:

In a multi-zone cluster, when configured with the proper placement constraints[0], HDFS logs indicate correct behaviour:

```
17/12/11 22:39:08 INFO hdfs.StateChange: BLOCK* registerDatanode: from DatanodeRegistration(10.10.0.253, datanodeUuid=7ef5c190-796a-412c-ac0a-33822953ce27, infoPort=9004, infoSecurePort=0, ipcPort=9005, storageInfo=lv=-56;cid=CID-692e6e53-7e99-4479-bbb6-a6b36a0509a5;nsid=1109145548;c=0) storage 7ef5c190-796a-412c-ac0a-33822953ce27
17/12/11 22:39:09 INFO net.NetworkTopology: Adding a new node: /USA-1/10.10.0.253:9003
17/12/11 22:39:09 INFO blockmanagement.BlockReportLeaseManager: Registered DN 7ef5c190-796a-412c-ac0a-33822953ce27 (10.10.0.253:9003).
17/12/11 22:39:10 INFO hdfs.StateChange: BLOCK* registerDatanode: from DatanodeRegistration(10.10.0.234, datanodeUuid=6eab0b2f-fdb0-4a0d-8c61-da3b0de0930a, infoPort=9004, infoSecurePort=0, ipcPort=9005, storageInfo=lv=-56;cid=CID-692e6e53-7e99-4479-bbb6-a6b36a0509a5;nsid=1109145548;c=0) storage 6eab0b2f-fdb0-4a0d-8c61-da3b0de0930a
17/12/11 22:39:10 INFO net.NetworkTopology: Adding a new node: /USA-2/10.10.0.234:9003
17/12/11 22:39:10 INFO blockmanagement.DatanodeManager: DN 10.10.0.234:9003 joining cluster has expanded a formerly single-rack cluster to be multi-rack. Re-checking all blocks for replication, since they should now be replicated cross-rack
17/12/11 22:39:16 INFO hdfs.StateChange: BLOCK* registerDatanode: from DatanodeRegistration(10.10.0.225, datanodeUuid=9abfd82a-9e1c-4abd-843e-37e8d33ceec2, infoPort=9004, infoSecurePort=0, ipcPort=9005, storageInfo=lv=-56;cid=CID-692e6e53-7e99-4479-bbb6-a6b36a0509a5;nsid=1109145548;c=0) storage 9abfd82a-9e1c-4abd-843e-37e8d33ceec2
17/12/11 22:39:16 INFO net.NetworkTopology: Adding a new node: /USA-3/10.10.0.225:9003
```

```
~$/bin/hdfs dfsadmin -printTopology
Rack: /USA-1
   10.10.0.253:9003 (10.10.0.253)

Rack: /USA-2
   10.10.0.234:9003 (10.10.0.234)

Rack: /USA-3
   10.10.0.225:9003 (10.10.0.225)
```

[0]:
```
 {
    "name_node": {
       "placement": "[[\"@zone\", \"GROUP_BY\"]]"
    },
    "data_node": {
       "placement": "[[\"@zone\", \"UNIQUE\"]]"
    }
  }
```

### CODE REVIEW

File: frameworks/hdfs/universe/resource.json [hunk 1]
```diff
@@ -8,7 +8,8 @@
       "hdfs-jre-tar-gz": "{{jre-jce-unlimited-url}}",
       "bootstrap-zip": "{{artifact-dir}}/bootstrap.zip",
       "scheduler-zip": "{{artifact-dir}}/hdfs-scheduler.zip",
-      "executor-zip": "{{artifact-dir}}/executor.zip"
+      "executor-zip": "{{artifact-dir}}/executor.zip",
+      "zone-resolver-sh": "http://downloads.mesosphere.com/hdfs/assets/zone-resolver.sh"
```
Comment (benclarkwood) [inline frameworks/hdfs/universe/resource.json]: Why not template this?

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/TaskUtils.java [hunk 2]
```diff
@@ -522,4 +525,41 @@ private static TaskInfo clearReservationIds(TaskInfo taskInfo) {
...
         return clearedResources;
     }
+
+    /**
+     * Determines if a task is launched in any zones.
+     * @param taskInfo The {@link TaskInfo} to get zone information from.
+     * @return A boolean indicating whether the task is in a zone.
+     */
+    public static boolean taskHasZone(TaskInfo taskInfo) {
+        Stream<Protos.Environment.Variable> envVars =
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/TaskUtils.java]: I'd just collapse this into a single instruction

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/TaskUtils.java [hunk 3]
```diff
@@ -522,4 +525,41 @@ private static TaskInfo clearReservationIds(TaskInfo taskInfo) {
...
+        return envVars.anyMatch(variable -> variable.getName().equals(EnvConstants.ZONE_TASKENV));
+    }
+
+    /**
+     * Gets the zone of a task.
+     * @param taskInfo The {@link TaskInfo} to get zone information from.
+     * @return A string indicating the zone the task is in.
+     */
+    public static String getTaskZone(TaskInfo taskInfo) {
+        Stream<Protos.Environment.Variable> envVars =
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/TaskUtils.java]: Ditto, collapse into a single instruction

File: frameworks/hdfs/universe/resource.json (hunk 1 above)
Comment (benclarkwood) [reply frameworks/hdfs/universe/resource.json]: Also, this script needs to be version controlled in this repo

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java [hunk 4]
```diff
@@ -170,6 +168,59 @@ public Response putFile(
         }
     }
 
+    /**
+     * Returns the Zone information for all of the tasks of the service.
+     */
+    @Path("/zone/tasks")
+    @GET
+    public Response getTaskNamesToZones() {
```
Comment (benclarkwood) [inline sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java]: Do all of these APIs get used? I don't see the `zone-resolver.sh` script anywhere in this PR?

File: frameworks/hdfs/universe/resource.json (hunk 1 above)
Comment (NimaVaziri) [reply frameworks/hdfs/universe/resource.json]: We _could_ put it as a template/config file, though there's not much to template in it.

**Comment by NimaVaziri** [comment]
@benclarkwood added the zone-resolver script

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java (hunk 4 above)
Comment (NimaVaziri) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java]: Nope, only the `getTaskIPsToZones` handler. Nick suggested we might as well expose a couple other endpoints to get zone info from, but these can be removed if you have a strong preference (I don't.)

File: frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/Main.java [hunk 5]
```diff
@@ -128,4 +122,56 @@ private static String getHDFSUserAuthMappings(Map<String, String> env, String en
         String authMappings = new String(hdfsUserAuthMappingsBytes, "UTF-8");
         return authMappings;
     }
+
+    private static PodSpec getJournalPodSpec(ServiceSpec serviceSpec) {
+        // Journal nodes avoid themselves and Name nodes.
+        PlacementRule placementRule = new AndRule(
+                TaskTypeRule.avoid(JOURNAL_POD_TYPE), TaskTypeRule.avoid(NAME_POD_TYPE));
+
+        if (getPodSpec(serviceSpec, JOURNAL_POD_TYPE).getPlacementRule().isPresent()) {
```
Comment (benclarkwood) [inline frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/Main.java]: You could refactor the last 2/3 or each of these into a common function if you'd like.

File: frameworks/hdfs/tools/zone-resolver.sh [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+export DATA_NODE_ZONE=$(curl $SCHEDULER_API_HOSTNAME/v1/state/zone/data/$1)
```
Comment (benclarkwood) [inline frameworks/hdfs/tools/zone-resolver.sh]: A comment or two in this file explaining what $1 is, the purpose of the script, etc would be lovely.

File: frameworks/hdfs/universe/marathon.json.mustache [hunk 7]
```diff
@@ -29,6 +29,7 @@
     "FRAMEWORK_NAME": "{{service.name}}",
     "FRAMEWORK_LOG_LEVEL": "{{service.log_level}}",
     "HDFS_VERSION": "hadoop-2.6.0-cdh5.11.0",
+    "TASKCFG_ALL_HDFS_VERSION": "hadoop-2.6.0-cdh5.11.0",
```
Comment (benclarkwood) [inline frameworks/hdfs/universe/marathon.json.mustache]: I'd drop the plain HDFS_VERSION one.

File: frameworks/hdfs/universe/marathon.json.mustache [hunk 8]
```diff
@@ -29,6 +29,7 @@
...
     "JRE_VERSION": "jre1.8.0_131",
```
Comment (benclarkwood) [inline frameworks/hdfs/universe/marathon.json.mustache]: This is out of date, which suggests to me that it isn't used anymore?

File: frameworks/hdfs/universe/resource.json (hunk 1 above)
Comment (benclarkwood) [reply frameworks/hdfs/universe/resource.json]: I meant, "why not abuse templating to copy over the simple script?"

Having it as a URI is fine of course.

File: sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java (hunk 4 above)
Comment (benclarkwood) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java]: 👍

Comment (benclarkwood) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java]: Are there tests for the other *Resource files? If yes then...

Comment (NimaVaziri) [reply sdk/scheduler/src/main/java/com/mesosphere/sdk/api/StateResource.java]: Yup, there are now

**Pull request merged**